internal/worker/objectstores3caller/worker.go

```diff
@@ -0,0 +1,273 @@
+// Copyright 2024 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package objectstores3caller
+
+import (
+	context "context"
+	"sync"
+	"time"
+
+	"github.com/juju/clock"
+	"github.com/juju/errors"
+	"github.com/juju/retry"
+	"github.com/juju/worker/v4"
+	"github.com/juju/worker/v4/catacomb"
+	"github.com/juju/worker/v4/dependency"
+
+	"github.com/juju/juju/controller"
+	"github.com/juju/juju/core/objectstore"
+	coretrace "github.com/juju/juju/core/trace"
+	"github.com/juju/juju/core/watcher"
+	"github.com/juju/juju/core/watcher/eventsource"
+	"github.com/juju/juju/internal/s3client"
+)
+
+const (
+	// States which report the state of the worker.
+	stateStarted       = "started"
+	stateClientUpdated = "client-updated"
+)
+
+const (
+	// default retry strategy for when the forbidden error is returned.
+	defaultRetryAttempts    = 10
+	defaultRetryDelay       = time.Second * 1
+	defaultRetryMaxDelay    = time.Second * 20
+	defaultRetryMaxDuration = time.Minute
+)
+
+// ControllerService is the interface that the worker uses to get the
+// controller configuration.
+type ControllerService interface {
+	// ControllerConfig returns the current controller configuration.
+	ControllerConfig(context.Context) (controller.Config, error)
+	// Watch returns a watcher that returns keys for any changes to controller
+	// config.
+	Watch() (watcher.StringsWatcher, error)
+}
+
+type workerConfig struct {
+	ControllerService ControllerService
+	HTTPClient        s3client.HTTPClient
+	NewClient         NewClientFunc
+	Logger            s3client.Logger
+	Clock             clock.Clock
+}
+
+// Validate returns an error if the workerConfig is not valid.
+func (cfg workerConfig) Validate() error {
+	if cfg.ControllerService == nil {
+		return errors.NotValidf("nil ControllerService")
+	}
+	if cfg.HTTPClient == nil {
+		return errors.NotValidf("nil HTTPClient")
+	}
+	if cfg.NewClient == nil {
+		return errors.NotValidf("nil NewClient")
+	}
+	if cfg.Clock == nil {
+		return errors.NotValidf("nil Clock")
+	}
+	return nil
+}
+
+type s3Worker struct {
+	internalStates chan string
+	catacomb       catacomb.Catacomb
+	config         workerConfig
+
+	mutex   sync.Mutex
+	session objectstore.Session
+}
+
+// NewWorker returns a new worker that wraps an S3 Session.
+func NewWorker(config workerConfig) (worker.Worker, error) {
+	return newWorker(config, nil)
+}
+
+func newWorker(config workerConfig, internalStates chan string) (*s3Worker, error) {
```

Comment:
Although not exported, this is a logical place to put some comments differentiating it from the unit S3 worker.